### PR TITLE
adding a function to forcibly update measurement infos

### LINKED ISSUE #2948: raw.add_channels  'buffer_size_sec' error
I'm trying to add a `stim` channel (a clean combination of multiple already existing stim channels) to a raw object:

```
# read raw
raw = read_raw_edf(fname, preload=True)

# clean stim
raw_stim = raw.pick_channels(chs_stim, copy=True)
data = clean_stim(raw_stim._data)

# append data
info = create_info(ch_names=['stim_clean'], sfreq=raw.info['sfreq'], ch_types=['stim'])
raw_stim = RawArray(data, info=info, first_samp=raw._first_samps[0])
raw.add_channels([raw_stim])
```

But I get the following error:

`RuntimeError: Don't know how to merge 'buffer_size_sec'. Make sure values are compatible.`

What shall I do?

cc @choldgraf @Eric89GXL

### PR DESCRIPTION
**Pull request by choldgraf**
In reference to #2948, this adds a function to forcible update parameters of target info objects using the values from a base info object. Generally meant to facilitate adding new channels that you created from RawArray. A couple thoughts worth mentioning:

* It hard-codes a few fields that it won't update `['chs', 'ch_names', and 'nchan']`. Is that OK and do you think I missed some?
* It will update the fields in target_info without copying first. I could do a deepcopy first and then return info objects, though that'll make it a little clunkier. I did make it so that using the force parameter in `add_channels` *will* deepcopy before updating the infos.
* In general, this is kind of a "dumb" implementation, so I tried to warn about potential ramifications in the doc for the "force" parameter in `add_channels`. I could also add more documentation to the _force_update... function itself.

WDYT? @Eric89GXL @kingjr @agramfort ?

EDIT by Eric89GXL:
Closes #2948.

### CODE REVIEW

**Comment by Eric89GXL** [comment]
It sounds okay to me, but I don't use the mentioned functions very often so hopefully other people have ideas. @jona-sassenhagen @teonbrooks do you use these functions much, and if so, would this help you and does it look reasonable?

**Comment by kingjr** [comment]
I don't have strong opinions or comments. I'm not playing with the `Info` object much.

**Comment by jona-sassenhagen** [comment]
No real input, I used add_channels once and that's it. It looks reasonable though.

**Comment by choldgraf** [comment]
I mostly use it when I want to display a stim channel that was recorded separately from my data. In my case this is pretty much always audio that I've tried to time-lock to the electrophysiological signal.

**Comment by choldgraf** [comment]
also I should say - whenever we call this from a user-facing function (add_channels is the only place that uses it currently), it will copy the Info object. It would only overwrite things if somebody called `_force_update_info` on their own.

File: mne/channels/channels.py [hunk 1]
```diff
@@ -613,7 +618,10 @@ def add_channels(self, add_list, copy=None):
 
         # Create final data / info objects
         data = np.concatenate(data, axis=con_axis)
-        infos = [self.info] + [inst.info for inst in add_list]
+        infos_to_add = deepcopy([inst.info for inst in add_list])
+        if force_update_info is True:
+            _force_update_info(self.info, infos_to_add)
+        infos = [self.info] + infos_to_add
         new_info = _merge_info(infos)
```
Comment (agramfort) [inline mne/channels/channels.py]: I would pass the force_update_info param to _merge_info and do the logic in there

File: mne/io/fiff/tests/test_raw_fiff.py [hunk 2]
```diff
@@ -1051,6 +1051,15 @@ def test_add_channels():
     )
     raw_new = raw_meg.copy().add_channels([raw_eeg])
 
+    # Testing force updates
+    raw_eeg_conf = deepcopy(raw_eeg)
```
Comment (agramfort) [inline mne/io/fiff/tests/test_raw_fiff.py]: deepcopy(raw_eeg) -> raw_eeg.copy()

**Comment by agramfort** [comment]
that's it for me now

File: mne/channels/channels.py (hunk 1 above)
Comment (choldgraf) [inline mne/channels/channels.py]: I thought about doing it that way but opted not just to keep from complicating that function more, but I'm happy to move this stuff in there. I can't really think of a case where you'd want to do the field updating *without* subsequently merging. It'd also remove an extra function...

Comment (agramfort) [inline mne/channels/channels.py]: to me it's better to keep the update logic in one place. You can keep
your _force_update_info
function but call it from _merge_info

Comment (choldgraf) [inline mne/channels/channels.py]: cool, will do

**Comment by choldgraf** [comment]
Mmk, I moved it within `merge_info`...that look OK?

File: mne/io/meas_info.py [hunk 3]
```diff
@@ -1477,3 +1484,18 @@ def _empty_info(sfreq):
...
     info._check_consistency()
     return info
+
+
+def _force_update_info(info_base, info_target):
+    """Update target info objects with values from info base."""
+    exclude_keys = ['chs', 'ch_names', 'nchan']
+    info_target = np.atleast_1d(info_target)
+    if not isinstance(info_base, Info):
+        raise ValueError('info_base must be of type Info')
```
Comment (agramfort) [inline mne/io/meas_info.py]: say what type you got

File: mne/io/meas_info.py [hunk 4]
```diff
@@ -1477,3 +1484,18 @@ def _empty_info(sfreq):
...
+    if not all([isinstance(ii, Info) for ii in info_target]):
+        raise ValueError('all target infos must both Info objects')
```
Comment (agramfort) [inline mne/io/meas_info.py]: sentence seems weird

**Comment by agramfort** [comment]
besides LGTM

**Comment by agramfort** [comment]
let me know when comments are addressed

**Comment by choldgraf** [comment]
OK, cleaned up the check + error message, how does that look?

File: mne/channels/channels.py [hunk 5]
```diff
@@ -560,14 +560,18 @@ def inst_has(attr):
         elif isinstance(self, Evoked):
             self.data = self.data.take(idx, axis=0)
 
-    def add_channels(self, add_list, copy=None):
+    def add_channels(self, add_list, force_update_info=False, copy=None):
```
Comment (Eric89GXL) [inline mne/channels/channels.py]: should probably put after `copy` even though it's about to be removed (could in 0.12 break someone's code)

File: mne/channels/channels.py [hunk 6]
```diff
@@ -560,14 +560,18 @@ def inst_has(attr):
...
 
         Parameters
         ----------
         add_list : list
             A list of objects to append to self. Must contain all the same
             type as the current object
+        force_update_info : bool
+            If True, force the info for objects to be appended to match the
+            values in `self`. This should generally only be used when adding
+            stim channels for which important metadata won't be overwritten.
```
Comment (Eric89GXL) [inline mne/channels/channels.py]: add `..versionadded:: 0.12` with newlines on either side to render properly (see existing examples)

File: mne/io/meas_info.py [hunk 7]
```diff
@@ -1477,3 +1484,31 @@ def _empty_info(sfreq):
...
+    info_target : mne.Info | list of mne.Info
+        The Info object(s) you wish to overwrite using info_base. These objects
+        will be modified in-place.
+    """
+    exclude_keys = ['chs', 'ch_names', 'nchan']
+    info_target = np.atleast_1d(info_target).ravel()
+    all_infos = np.hstack([info_base, info_target])
+    msg = 'Inputs must be of type Info. Found type {0}'
+    if not all([isinstance(ii, Info) for ii in all_infos]):
+        raise ValueError(msg.format(type(ii)))
```
Comment (Eric89GXL) [inline mne/io/meas_info.py]: Please follow our usual pattern of string with %s, and don't make it a separate var

**Comment by Eric89GXL** [comment]
Can you add a test that covers the use case that actually motivated the change? It's good to include that so we see the value and don't break it in the future.

File: mne/channels/channels.py (hunk 5 above)
Comment (choldgraf) [inline mne/channels/channels.py]: cool, done (going through your suggestions now)

File: mne/channels/channels.py (hunk 6 above)
Comment (choldgraf) [inline mne/channels/channels.py]: done

File: mne/io/meas_info.py (hunk 7 above)
Comment (choldgraf) [inline mne/io/meas_info.py]: ok

**Comment by choldgraf** [comment]
ok, addressed comments and updated the test so that it recreates the issue of using RawArray then adding channels w/ that. Is that OK?

**Comment by agramfort** [comment]
I am happy if @Eric89GXL is happy

File: mne/io/fiff/tests/test_raw_fiff.py [hunk 8]
```diff
@@ -1056,6 +1056,18 @@ def test_add_channels():
...
     assert_true(all(ch not in raw_new.ch_names for ch in raw_stim.ch_names))
 
+    # Testing force updates
+    raw_arr_info = create_info(['1', '2'], raw_meg.info['sfreq'], 'eeg')
+    orig_head_t = raw_arr_info['dev_head_t']
+    raw_arr = np.random.randn(2, raw_eeg.n_times)
+    raw_arr = RawArray(raw_arr, raw_arr_info)
+    # This should error because of conflicts in Info
+    assert_raises(ValueError, raw_meg.copy().add_channels, [raw_arr])
+    raw_conf = raw_meg.copy().add_channels([raw_arr],
```
Comment (Eric89GXL) [inline mne/io/fiff/tests/test_raw_fiff.py]: this is going to be a `flake8` error because `raw_conf` is never used

Comment (choldgraf) [inline mne/io/fiff/tests/test_raw_fiff.py]: ah good point, I just want to make sure that the function won't throw an error. I updated it to throw away the output, will that work?

File: mne/io/fiff/tests/test_raw_fiff.py [hunk 9]
```diff
@@ -1056,6 +1056,17 @@ def test_add_channels():
...
     assert_true(all(ch not in raw_new.ch_names for ch in raw_stim.ch_names))
 
+    # Testing force updates
+    raw_arr_info = create_info(['1', '2'], raw_meg.info['sfreq'], 'eeg')
+    orig_head_t = raw_arr_info['dev_head_t']
+    raw_arr = np.random.randn(2, raw_eeg.n_times)
+    raw_arr = RawArray(raw_arr, raw_arr_info)
+    # This should error because of conflicts in Info
+    assert_raises(ValueError, raw_meg.copy().add_channels, [raw_arr])
+    _ = raw_meg.copy().add_channels([raw_arr], force_update_info=True)
```
Comment (Eric89GXL) [inline mne/io/fiff/tests/test_raw_fiff.py]: FYI you don't actually need the `_ =`, but this is fine

Comment (choldgraf) [inline mne/io/fiff/tests/test_raw_fiff.py]: strangely enough only of the builds errored with Flakes on this line, though the others didn't...I changed it so there's no variable assigned to the output.

**Comment by Eric89GXL** [comment]
Only one build runs flake8

**Comment by choldgraf** [comment]
I suppose that explains it then :)

**Comment by Eric89GXL** [comment]
Thanks @choldgraf

**Comment by choldgraf** [comment]
np!

**Pull request merged**